NXBT-3255: Configure Nexus as docker internal registry (#51)
nuxeo/jx-platform-env: docs/platform-setup.md

> # Setup Of The Platform Jenkins X Team

Reviewer: Sorry for being super picky but usually, and according to https://capitalizemytitle.com/, we don't capitalize "of", "the", ...

Author: you are wright

Author: changes applied

## Disable GitHub Multibranch Status

### Prerequisite

- Jenkins is installed.
- The [Disable GitHub Multibranch Status](https://plugins.jenkins.io/disable-github-multibranch-status/) plugin is installed.

Open the platform team's [Jenkins](https://jenkins.platform.dev.nuxeo.com/).

> 

Author: Hello, 

I have separated  the two sections 'Disable GitHub Multibranch Status' and 'Configure  Nexus Docker Registry'

- Select "Disable GitHub Notifications".

- Click on "Save".

> ## Deploy Nexus Docker Registry

Author: will be taken to account

### Prerequisite

- Nexus is installed.
- NGINX Ingress Controller is installed.

Deploying docker registry with nexus should follow this steps:

### Create New Docker Repository

> Open the platform team's [Nexus](https://nexus.platform.dev.nuxeo.com/). Start new session using admin user

Author: changes applied

> #### Blop Storage And Repository Configuration

Author: will be taken to account

- Create a blob storage : `docker-registry`

  - In Server Administration and Configuration > Blob Stores,  click on the "Create blob store" button

  - Complete all required fields and click on "Save"

- Create a docker (hosted) nexus repository: `docker-registry`

  - In Server Administration and Configuration > Repositories,  click on the "Create repository" button

>   - Complete all required fields and click on "Save"

Reviewer 2: Shouldn't we mention at some point to use the previously created blob store?

Author: will be taken to account

### Cleanup Policy Configuration

> For each nexus docker registry, create a new policy cleanup and  apply to it.

Reviewer 2: extra space between _and_ and _apply_.

Reviewer: Here we understand that we need to create 1 policy for each docker repository!

If I'm not mistaken, the cleanup policy needs to be created only once for the `docker` format.
Then indeed, it needs to be applied for each docker repository.

Author: sorry for the misunderstanding

#### Create New Cleanup Policy

> - In Server Administration and Configuration > Cleanup Policies,  click on the "Create Cleanup Policy" button

Reviewer 2: extra space between _Policies,_ and _click_

> - To keep all nuxeo builder and jenkins master images ( latest tag for example ) add this asset name matcher `~(.*nuxeo/builder-.*/manifests/latest|.*nuxeo/platform.*-jenkinsx/manifests/[0-9]{1,}.[0-9]{1,}.[0-9]{1,})`

Reviewer 2: Maybe:
```
(latest tag for example)
```
?

- Complete all required fields and click on "Save"

#### Apply Cleanup Policy

- Open an existing nexus repository configuration for instance [docker-registry](https://nexus.platform.dev.nuxeo.com/#admin/repository/repositories:docker-registry)

- Go done to cleanup section and select the corresponding item

- Click on "Save".

### Create New Docker Ingress Service

This ingress service will act as Reverse Proxy for nexus docker repository.

> The reverse proxy would direct docker commands ( pull, push or login ) received at

Reviewer 2: Maybe:
```
(pull, push or login)
```
?

> `https://docker.platform.dev.nuxeo.com`

Reviewer 2: Currently this is `https://docker-registry.platform.dev.nuxeo.com`, shouldn't we want to use the same?

Reviewer: `docker-registry` refers to the [Docker registry Helm chart](https://github.com/helm/charts/tree/master/stable/docker-registry) installed by default with the Jenkins X platform.
Since we're disabling it, it doesn't seem a problem to me to use `docker.xxx` directly.
Just like `docker.packages.nuxeo.com`.

Reviewer: Also, maybe remove the backticks: though it will add a Markdown linter warning, we might want those URLs' to be clickable.

to

`https://nexus.platform.dev.nuxeo.com/repository/docker-registry/`

> Create an ingress configuration file following this [example](../templates/docker-ingress.json)

Reviewer 2: Why a json and not a yaml file?

Author: I am more comfortable with JSON. But if 'yaml is more appropriate I can change the file type.

Reviewer 2: I would use the same format everywhere?

Reviewer: Yes, we tend to use YAML everywhere for Kubernetes resources.

Reviewer: Also, maybe use bullet points for each steps?
```
- Create an ingress configuration...
...
- Create new ingress service from the configuration file
```

Author: changes applied

Create new ingress service from the configuration file

```bash
kubectl create -f templates/docker-ingress.json
```

Check if the ingress service is ready

```bash
kubectl get -f templates/docker-ingress.json
```

List the catalog from the new docker registry under nexus

```bash
DOCKER_REGISTRY=$(kubectl get -f templates/docker-ingress.json -o json --output=jsonpath={.spec.rules[].host} )
curl https://${DOCKER_REGISTRY}/v2/_catalog
```

> ### Update The Config.json Secret

Reviewer 2: Should we use lowercase for config.json?

Create/update Docker auth for a given host(docker.platform.dev.nuxeo.com for example) and user in the config.json file

```bash
export NAMESPACE=platform
NEXUS_PASSWORD=$(kubectl get secret nexus -n ${NAMESPACE} -o=jsonpath='{.data.password}' | base64 --decode)
jx create docker auth --host "docker.${NAMESPACE}.dev.nuxeo.com" --user "admin" --secret "${NEXUS_PASSWORD}"
```

If building your images with Kaniko, you have to add the Kubernetes secret that contains the config.json Docker configuration (in all existing skaffold.yaml)

```yaml
cluster:
    ...
    dockerConfig:
      secretName: jenkins-docker-cfg

> ```

Reviewer 2: dockerConfig and secretName are too much indented to the right.

Author: will be taken to account

### Update PodTemplates

#### Create A Secret Based On Existing Docker Credentials

> We suppose you have a json file containing credentials for `docker.platform.dev.nuxeo.com`

Reviewer 2: We should have it no? as it is the previous step.

Author: The previous if i'm not wrong  define the secret to be used by Kaniko.
In this section we are defining  the secret to be used by kubernetes to retrieve pod template image.

we will have two secret : 
 - jenkins-docker-cfg ( Opaque Secret used by Kaniko)
 - nexus-docker-cfg ( Generic Secret with kubernetes.io/dockerconfigjson type used by kubernetes to pull pod template image )

because they store the docker credentials with different  methods .
 - Kubernetes save the docker credential  on section '.dockerconfigjson'
 - Kaniko save the docker credential  on section 'config.json'

Reviewer 2: Maybe it's worth adding a comment explaining this.

Reviewer: Yes, definitely.

Author: changes applied

- Select "Disable GitHub Notifications".

- Click on "Save".

## Configure Nexus Docker Registry

### Prerequisite

- Nexus is installed.

> - NGINX Ingress Controller is installed.

Reviewer: Is this the `jxing-nginx-ingress-controller-846d77d58b-j4vm9` pod of the `default` namespace ?
Maybe just add "in the `default` namespace"

Author: changes applied

> Configuring docker registry with nexus should follow this steps:

Reviewer: I think you can remove this sentence, it is clear that we need to follow each step.

Author: changes applied

### Create New Docker Repository

Open the platform team's [Nexus](https://nexus.platform.dev.nuxeo.com/). Start new session using admin user

> #### Blob Storage And Repository Configuration

Reviewer: This is the only subsection in `Create New Docker Repository`.
Maybe have these 2 subsections instead:
```

#### Blob Storage

#### Repository
```

Author: changes applied

- Create a blob storage : `docker-registry`

>   - In Server Administration and Configuration > Blob Stores,  click on the "Create blob store" button

Reviewer: ```suggestion
  - In Server Administration and Configuration > Repository > Blob Stores,  click on the "Create blob store" button
```

Author: Hi,

Blob storage creation button is not under Repository ???

Author: changes applied

  - Complete all required fields and click on "Save"

- Create a docker (hosted) nexus repository: `docker-registry`

>   - In Server Administration and Configuration > Repositories,  click on the "Create repository" button

Reviewer: ```suggestion
  - In Server Administration and Configuration > Repository > Repositories,  click on the "Create repository" button
```

Author: changes applied

  - On storage section, use the previously created blob store

  - Complete all required fields and click on "Save"

### Cleanup Policy Configuration

For each nexus docker registry, create a new policy cleanup and apply to it.

#### Create New Cleanup Policy

> - In Server Administration and Configuration > Cleanup Policies,  click on the "Create Cleanup Policy" button

Reviewer: ```suggestion
- In Server Administration and Configuration > Repository > Cleanup Policies,  click on the "Create Cleanup Policy" button
```

Author: changes applied

- To keep all nuxeo builder and jenkins master images ( latest tag for example ) add this asset name matcher `~(.*nuxeo/builder-.*/manifests/latest|.*nuxeo/platform.*-jenkinsx/manifests/[0-9]{1,}.[0-9]{1,}.[0-9]{1,})`

- Complete all required fields and click on "Save"

#### Apply Cleanup Policy

> - Open an existing nexus repository configuration for instance [docker-registry](https://nexus.platform.dev.nuxeo.com/#admin/repository/repositories:docker-registry)

Reviewer: I would say: "Open the [docker-registry](https://nexus.platform.dev.nuxeo.com/#admin/repository/repositories:docker-registry) repository configuration for instance."

Author: changes applied

> - Go done to cleanup section and select the corresponding item

Reviewer: ```suggestion
- Scroll down to the Cleanup Policies section and select the corresponding item
```

- Click on "Save".

### Create New Docker Ingress Service

> This ingress service will act as Reverse Proxy for nexus docker repository.

Reviewer: ```suggestion
This ingress service will act as reverse proxy for the nexus docker repository.
```

> The reverse proxy would direct docker commands ( pull, push or login ) received at

Reviewer: ```suggestion
The reverse proxy will redirect docker commands (pull, push or login) received at:
```

Author: changes applied

`https://docker.platform.dev.nuxeo.com`

to

`https://nexus.platform.dev.nuxeo.com/repository/docker-registry/`

Create an ingress configuration file following this [example](../templates/docker-ingress.json)

Create new ingress service from the configuration file

```bash
kubectl create -f templates/docker-ingress.json
```

Check if the ingress service is ready

```bash
kubectl get -f templates/docker-ingress.json
```

> List the catalog from the new docker registry under nexus

Reviewer: ```suggestion
List the catalog from the new nexus docker registry
```

Author: changes applied

```bash
DOCKER_REGISTRY=$(kubectl get -f templates/docker-ingress.json -o json --output=jsonpath={.spec.rules[].host} )
curl https://${DOCKER_REGISTRY}/v2/_catalog
```

> ### Update The Config.json Secret

Reviewer: Maybe we could use something more functional here, instead of "Update The Config.json Secret", "Update the Jenkins X Docker Configuration".

And, as an introduction in this section, we could add a small paragraph explaining that we need to add authentication for the newly created Docker registry to allow pushing images.

Author: changes applied

> Create/update Docker auth for a given host(docker.platform.dev.nuxeo.com for example) and user in the config.json file

Reviewer: ```suggestion
Create/update Docker authentication for the newly added [Docker registry](docker.platform.dev.nuxeo.com) in the Jenkins X Docker configuration secret (`jenkins-docker-cfg`).

```

Author: changes applied

```bash

> export NAMESPACE=platform

Reviewer: I think `export` is not needed.

NEXUS_PASSWORD=$(kubectl get secret nexus -n ${NAMESPACE} -o=jsonpath='{.data.password}' | base64 --decode)
jx create docker auth --host "docker.${NAMESPACE}.dev.nuxeo.com" --user "admin" --secret "${NEXUS_PASSWORD}"
```

> If building your images with Kaniko, you have to add the Kubernetes secret that contains the config.json Docker configuration (in all existing skaffold.yaml)

Reviewer: Since we
```suggestion
Since we are building our images with Kaniko, you have to add the Kubernetes secret that contains the config.json Docker configuration (in all existing skaffold.yaml)
```

> If building your images with Kaniko, you have to add the Kubernetes secret that contains the config.json Docker configuration (in all existing skaffold.yaml)

Reviewer: ```suggestion
If building your images with Kaniko, you have to specify the Kubernetes secret that contains the config.json Docker configuration in all the skaffold.yaml files.
```

Author: changes applied

```yaml
cluster:
  ...
  dockerConfig:
    secretName: jenkins-docker-cfg
```

> ### Update PodTemplates

Reviewer: Same remark as for the previous section, maybe something more functional like "Handle authentication for Pulling Images".

And add a small paragraph explaining that we need to add authentication for the newly created Docker registry to allow **pulling** images.

Author: changes applied

#### Create A Secret Based On Existing Docker Credentials

We suppose you have a json file containing credentials for `docker.platform.dev.nuxeo.com`

```bash
kubectl create secret generic nexus-docker-cfg \
    --from-file=.dockerconfigjson=<PATH>/config.json \
    --type=kubernetes.io/dockerconfigjson
```

> #### Add ImagePullSecret

Reviewer: Adding `ImagePullSecret` is the implementation, prefer something more functional such as "Configure Pod Templates".

> If a PodTemplates need to retrieve from nexus docker repository, this line should be added

Reviewer: ```suggestion
If a PodTemplate needs to retrieve an image from the nexus docker repository, this line must be added:
```

Author: changes applied

Author: changes applied

- Select "Disable GitHub Notifications".

- Click on "Save".

## Configure Nexus Docker Registry

### Prerequisite

- Nexus is installed.
- NGINX Ingress Controller is installed in the default namespace.

### Create New Docker Repository

Open the [Nexus Repository Manager](https://nexus.platform.dev.nuxeo.com/). Start new session using admin user

#### Blob Storage

Create a blob storage : `docker-registry`

- In Server Administration and Configuration > Repository > Blob Stores, click on the "Create blob store" button

- Complete all required fields and click on "Save"

#### Repository

Create a docker (hosted) nexus repository: `docker-registry`

- In Server Administration and Configuration > Repository > Repositories, click on the "Create repository" button

> - On storage section, use the previously created blob store

Reviewer: ```suggestion
- In the Storage section, use the previously created blob store
```

- Complete all required fields and click on "Save"

### Cleanup Policy Configuration

Create a new policy cleanup and apply it for each nexus docker registry you have created if needed.

#### Create New Cleanup Policy

- In Server Administration and Configuration > Repository > Cleanup Policies, click on the "Create Cleanup Policy" button

- To keep all nuxeo builder and jenkins master images (latest tag for example) add this asset name matcher `~(.*nuxeo/builder-.*/manifests/latest|.*nuxeo/platform.*-jenkinsx/manifests/[0-9]{1,}.[0-9]{1,}.[0-9]{1,})`

- Complete all required fields and click on "Save"

#### Apply Cleanup Policy

- Open the [docker-registry](https://nexus.platform.dev.nuxeo.com/#admin/repository/repositories:docker-registry) repository configuration for instance

- Scroll down to the Cleanup Policies section and select the corresponding item

- Click on "Save".

### Create New Docker Ingress Service

This ingress service will act as reverse proxy for the nexus docker repository.

> The reverse proxy will direct docker commands (pull, push or login) received at:

Reviewer: ```suggestion
The reverse proxy will redirect docker commands (pull, push or login) received at:
```

```

### Update the Jenkins X Docker Configuration

Create/update Docker authentication for the newly added [Docker registry](docker.platform.dev.nuxeo.com) in the Jenkins X Docker configuration secret (`jenkins-docker-cfg`).

```bash
NAMESPACE=platform
NEXUS_PASSWORD=$(kubectl get secret nexus -n ${NAMESPACE} -o=jsonpath='{.data.password}' | base64 --decode)
jx create docker auth --host "docker.${NAMESPACE}.dev.nuxeo.com" --user "admin" --secret "${NEXUS_PASSWORD}"
```

If building your images with Kaniko, you have to specify the Kubernetes secret that contains the config.json Docker configuration in all the skaffold.yaml files.

```yaml
cluster:
  ...
  dockerConfig:
    secretName: jenkins-docker-cfg
```

### Handle Authentication for Pulling Images

An authentication should be configured for the newly created docker registry to enable pushing or pulling images.

#### Create a Secret Based on Existing Docker Credentials

We suppose you have a json file containing credentials for `docker.platform.dev.nuxeo.com`

```bash
kubectl create secret generic nexus-docker-cfg \
    --from-file=.dockerconfigjson=<PATH>/config.json \
    --type=kubernetes.io/dockerconfigjson
```

> unfortunately nexus-docker-cfg and jenkins-docker-cfg can seem to be duplicated. it is due to secret format.

Reviewer: ```suggestion
Unfortunately, `nexus-docker-cfg` and `jenkins-docker-cfg` can seem duplicated, it is due to secret format:
```

> - Kubernetes expect to have secret with `kubernetes.io/dockerconfigjson`  type and data to be stored on `data/.dockerconfigjson`  section.

Reviewer: ```suggestion
- Kubernetes expects to have a secret with the `kubernetes.io/dockerconfigjson` type and data to be stored in `data/.dockerconfigjson`.
```

> - For Kaniko the expected secret  type is `Opaque` and data to be stored on `data/config.json` section.

Reviewer: ```suggestion
- Kaniko expects to have a secret with the `Opaque` type and data to be stored in `data/config.json`.
```
